lib/trln_argon/solr_document/citation.rb

```diff
@@ -1,53 +1,166 @@
+require 'citeproc'
+require 'csl/styles'
+require 'bibtex'
+require 'citeproc/ruby'
+require 'json'
+
 module TrlnArgon
   module SolrDocument
     module Citation
-      def citable?
-        oclc_number.present? && citations.present?
-      end
+      MAPPING = {
+        'id' => 'id',
+        'title_main' => 'title',
+        'publication_year_sort' => 'year',
+        'names_a' => 'author',
+        'note_general_a' => 'note',
+        'isbn_number_a' => 'number',
+        'physical_description_a' => 'pages',
+        'publisher_a' => 'publisher',
+        'resource_type_a' => 'type',
+        'edition_a' => 'edition',
+        'publisher_location_a' => 'address',
+        'language_a' => 'language',
+        'issn_primary_a' => 'issn',
+        'url_a' => 'url'
+      }.freeze
 
       def citations
-        @all_citations ||= begin
-          citation_hash = {}
-          desired_formats.each do |format|
-            citation = world_cat_api_response(format).html_safe
-            if citation != 'info:srw/diagnostic/1/65Record does not exist'
-              citation_hash[format] = citation
-            end
+        @all_citations ||= generate_citations
+      end
+
+      def generate_citations
+        citation_hash = {}
+        result = map_document_values
+        bibtex_string = convert_to_bibtex_format(result)
+        bibliography = BibTeX.parse(bibtex_string)
+        process_citations(bibliography, citation_hash)
+        citation_hash
+      end
+
+      private
+
+      def map_document_values
+        MAPPING.each_with_object([]) do |(key, value), output|
+          docvalue = fetch(key, [])
+          next if docvalue.empty?
+
+          case key
+          when 'url_a'
+            process_url(docvalue, value, output)
+          when 'names_a'
+            process_names(docvalue, output)
+          else
+            process_docvalue(output, value, docvalue)
           end
-          citation_hash
         end
       end
 
-      private
+      def process_url(docvalue, value, output)
+        processed_value = parse_url(docvalue)
+        process_docvalue(output, value, processed_value)
+      end
 
-      def world_cat_api_response(format)
-        url = world_cat_uri(format)
+      def process_names(docvalue, output)
+        authors, editors = parse_authors_and_editors(docvalue)
+        process_docvalue(output, 'author', authors) unless authors.empty?
+        process_docvalue(output, 'editor', editors) unless editors.empty?
+      end
+
+      def parse_url(docvalue)
+        JSON.parse(docvalue.first)['href']
```

Comment:
We'll need to parse the `url_a` values more than this. Just getting the first `href` is going to result in a lot of unwanted URLs getting presented in the citations, especially in cases where the URLs are `type: related` (where they'll point to some other resource like a finding aid) or `type: other` (where they might lead to e.g., a Table of Contents). We'll also have cases like periodicals where there are multiple URLs and only including the first will be incorrect.